ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT/ChainState.hs

```diff
@@ -35,103 +41,173 @@ import           Codec.Serialise.Decoding (Decoder)
-      -- > G      | 3 (G)        | [B1, B2, B3]
-      -- > G      | 4 (G)        | [B1, B2, B3, B4]
-      -- > S1     | 4 (B1)       | [B2, B3, B4, B5]
-      -- > S2     | 4 (B2)       | [B3, B4, B5, B6]
-      -- > S3     | 4 (B3)       | [B4, B5, B6, B7]
+      -- We should have precisely @k@ signatures after the anchor, unless
       --
-      -- At this point the anchor becomes equal to the (slot number of) the
-      -- oldest block that we can roll back to.
-      anchor  :: !(WithOrigin SlotNo)
+      -- 1. We are near genesis, or
+      -- 2. After a during (during a switch-to-fork)
+    , postAnchor :: !(StrictSeq (PBftSigner c))
 
-      -- | Sequence in increasing order of slots and corresponding genesis keys
-    , signers :: !(StrictSeq (PBftSigner c))
+      -- | Signatures before the window
+    , preWindow  :: !(StrictSeq (PBftSigner c))
 
-      -- | Count for each genesis key
+      -- | Signatures in the window
       --
-      -- Invariant:
-      --
-      -- > counts == fromSlots signers
-      --
-      -- We maintain it as an explicit value for improved performance.
-    , counts  :: !(Map (PBftVerKeyHash c) Int)
+      -- We should have precisely @n@ signatures the window, unless we are
+      -- near genesis.
+    , inWindow   :: !(StrictSeq (PBftSigner c))
+
+      -- | Cached counts of the signatures in the window
+    , counts     :: !(Map (PBftVerKeyHash c) Word64)
     }
   deriving (Generic)
 
+{-------------------------------------------------------------------------------
+  Invariant
+-------------------------------------------------------------------------------}
+
+size :: Num b => StrictSeq a -> b
+size = fromIntegral . Seq.length
+
+-- | Re-compute cached counts
+computeCounts :: PBftCrypto c
+              => StrictSeq (PBftSigner c)  -> Map (PBftVerKeyHash c) Word64
+computeCounts inWindow =
+    repeatedly (incrementKey . pbftSignerGenesisKey)
+               (Foldable.toList inWindow)
+               Map.empty
+
+invariant :: PBftCrypto c
+          => SecurityParam -> WindowSize -> PBftChainState c -> Either String ()
+invariant (SecurityParam k)
+          (WindowSize n)
+          st@PBftChainState{..}
+        = runExcept $ do
+    unless (size postAnchor <= k) $
+      failure "Too many post-anchor signatures"
+
+    unless (size preAnchor <= n) $
+      failure "Too many pre-anchor signatures"
+
+    unless (size inWindow <= n) $
+      failure "Too many in-window signatures"
+
+    unless (size preWindow <= k) $
+      failure "Too many pre-window signatures"
+
+    unless (size preWindow + size inWindow <= k + n) $
+      failure "Too many signatures"
+
+    unless (preAnchor <> postAnchor == preWindow <> inWindow) $
+      failure "Inconsistent signature split"
+
+    unless (computeCounts inWindow == counts) $
+      failure "Cached counts incorrect"
+  where
+    failure :: String -> Except String ()
+    failure err = throwError $ err ++ ": " ++ show st
+
+-- TODO: Make this a CPP flag
```

Comment:
Docstring: the tests for `PBftChainState` already check the invariant, so they don't rely on this; but would be nice to enable it for the other consensus tests, just as a sanity check (maybe as part of https://github.com/input-output-hk/ouroboros-network/pull/1248).